keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,296 @@
+
+#### Task Application
+
+Once KUDO has assembled a full set of templates for a task, they will be applied using Kubernetes server-side apply at the same time.
+
+### Parameters
+
+Parameters are a map of Parameter structs, with the key being the name of the parameter. A default value can be set inside of a parameter, as well as a description. When no default is specified, the value is required. If an Instance is created without a required parameter, it will have an Error event added to it describing the validation failure. As described in [Execution State](#execution-state), they are wrapped into the `.Params` object for use by all templated objects. All keys, even arbitrary keys unused by KUDO, are wrapped into the `.Params` object for potential use by other templates.
+
+Parameters are intentionally open for extension for adding fields in the future for validation and more.
+
+### Templates
+
+A template is a standard Kubernetes manifest which **MAY** have additional Go Templating. These Go Templates include [Sprig](https://github.com/masterminds/sprig) to provide a standard library of functions familiar to Helm users. Additional values are available and are described in [Execution State](#execution-state).
+
+## Extensions and Bases
+
+Extensions and bases describe a mechanism for building operators or extensions to operators from a given base. As an example, a base can be an operator, a Helm chart, a CNAB bundle, or any future format that describes the deployment of a set of resources.
+
+In this document an **extension** is any KUDO operator that extends from some base. A **base** is the complete set of manifests, metadata, and other files provided by that base's type. A base should provide the complete information that users of that base tool are expected to have. The base types and what they expose to charts that extend from them are described in their respective sub-sections.
+
+To support extending from a base, `operator.yaml` is extended to support the `extends` keyword:
+
+```yaml
+extends:
+  kudo:
+    operator: "mysql"
+    version: "5.7"
+```
+
+After extending, the base resources are inherited by the extending operator. The behavior of extensions and values available from the base are described in their corresponding sub-section.
+
+When a task is defined in an extended operator, it **replaces** the task from the base. The tasks available are dependent on the base type and are described more in detail in their corresponding sub-section. To support extensibility of tasks that have been replaced, a few new keywords are available in `operator.yaml` to support further control over extensions:
+
+When a plan is defined in an extended operator, it **replaces** the plan from the base. The plans available are dependent on the base type and are described more in detail in their corresponding sub-section.
+
+### Task Extensions
```

Comment:
If we have `Task Extensions` and `Plan Extensions` there is no need for `Step Extensions` ?

Reply:
Exactly. You just replace a plan with new steps in that case, since steps are really just a list of tasks.

Reply:
maybe we should note that ^^^ in the doc in case others have the same question?

Reply:
I still feel like copying an entire Step instead of duplicating it with all its Tasks could make sense in some cases but I see what you mean. Sgtm.

Overall, somehow I am having a hard time to fully grasp how this will look like implemented (thats maybe also the reason for my question above). How would an example look like?